File: android/app/src/main/java/com/cochat/app/data/repository/ChatsRepository.kt
package com.cochat.app.data.repository

import com.cochat.app.data.model.ChatSummary
import com.cochat.app.data.model.Message
import com.cochat.app.data.remote.ApiService
import com.cochat.app.data.remote.OpenPrivateChatRequest
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File

class ChatsRepository(private val api: ApiService) {

    suspend fun list(): List<ChatSummary> = api.listChats().chats

    suspend fun openPrivate(userId: String): ChatSummary = api.openPrivateChat(OpenPrivateChatRequest(userId)).chat

    suspend fun messages(chatId: String, before: String? = null): List<Message> =
        api.getMessages(chatId, before).messages

    suspend fun sendMessage(chatId: String, text: String?, file: File?, mimeType: String?): Message {
        val parts = mutableMapOf<String, okhttp3.RequestBody>()
        if (!text.isNullOrBlank()) {
            parts["text"] = text.toRequestBody("text/plain".toMediaType())
        }
        val filePart = file?.let {
            val body = it.asRequestBody((mimeType ?: "application/octet-stream").toMediaType())
            MultipartBody.Part.createFormData("file", it.name, body)
        }
        return api.sendMessage(chatId, parts, filePart).message
    }

    suspend fun markRead(chatId: String) {
        api.markChatRead(chatId)
    }
}

File: ios/CoChat/Repositories/ChatsRepository.swift
import Foundation

final class ChatsRepository {
    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func list() async throws -> [ChatSummary] {
        let res: ChatsEnvelope = try await api.get("chats")
        return res.chats
    }

    func openPrivate(userId: String) async throws -> ChatSummary {
        let res: ChatEnvelope = try await api.post("chats/private", body: OpenPrivateChatRequest(userId: userId))
        return res.chat
    }

    func messages(chatId: String, before: String? = nil) async throws -> [Message] {
        var query: [String: String] = [:]
        if let before { query["before"] = before }
        let res: MessagesEnvelope = try await api.get("chats/\(chatId)/messages", query: query)
        return res.messages
    }

    func sendMessage(chatId: String, text: String?, fileData: Data?, fileName: String?, mimeType: String?) async throws -> Message {
        var fields: [String: String] = [:]
        if let text, !text.isEmpty { fields["text"] = text }
        let res: MessageEnvelope = try await api.postMultipart(
            "chats/\(chatId)/messages",
            fields: fields,
            fileField: fileData != nil ? "file" : nil,
            fileData: fileData, fileName: fileName, mimeType: mimeType
        )
        return res.message
    }

    func markRead(chatId: String) async throws {
        try await api.postVoid("chats/\(chatId)/read")
    }
}
